analyzers/src/SonarAnalyzer.Common/AnalysisContext/SonarCompilationStartAnalysisContext.cs

```diff
@@ -38,6 +38,26 @@ public sealed class SonarCompilationStartAnalysisContext : SonarAnalysisContextB
         Context.RegisterSemanticModelAction(x => action(new(AnalysisContext, x)));
 
     public void RegisterNodeAction<TSyntaxKind>(GeneratedCodeRecognizer generatedCodeRecognizer, Action<SonarSyntaxNodeReportingContext> action, params TSyntaxKind[] syntaxKinds)
-        where TSyntaxKind : struct =>
-        AnalysisContext.RegisterNodeAction(generatedCodeRecognizer, action, syntaxKinds);
+        where TSyntaxKind : struct
+    {
+        if (HasMatchingScope(AnalysisContext.SupportedDiagnostics))
+        {
+            Context.RegisterSyntaxNodeAction(x =>
+                    Execute<SonarSyntaxNodeReportingContext, SyntaxNodeAnalysisContext>(
+                        new(AnalysisContext, x), action, x.Node.SyntaxTree, generatedCodeRecognizer),
+                syntaxKinds);
+        }
+    }
+
+    /// <param name="sourceTree">Tree that is definitely known to be analyzed. Pass 'null' if the context doesn't know a specific tree to be analyzed, like a CompilationContext.</param>
```

Comment:
```suggestion
    /// <inheritdoc cref="SonarAnalysisContext.Execute" />
```

Reply:
Add a comment about SonarAnalysisContext.Execute being removed and replaced by this method in the long run.